[Operator Versioning][Edge] Use check in cpp source file for upgrader

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67731
* __->__ #67728



1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

Commit:
[Operator Versioning][Edge] Use check in cpp source file for upgrader

1. Check in upgrader_mobile.h and it will be replaced by code-gen from yaml file in next few pr. Let me know if this file makes sense.
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

[ghstack-poisoned]

iseeyuan commented on torch/csrc/jit/mobile/upgrader_mobile.h:
```diff
@@ -0,0 +1,70 @@
+    {
+        // The following are called from setup sections.
+        {"div_Tensor_0_3",
+         to_tuple(std::vector<c10::IValue>{
```
It's quite space consuming. Could you directly build a mobile::Function here, instead of those strings and later parsed to a function?

cccclai replied:
Address in the latest commit.

iseeyuan commented:
Thanks! Two thoughts:
1. The overhead of the upgraders. I don't think we need to put the literal bytecode strings in the header file and parse it at loading time. It may not be efficient for both time and size. For instructions, constants, etc., we could directly create the in-memory lists, without any strings or ivalues of tuples. Operators may be the only exception (We could only resolve the op kernel at runtime). 
2. The entire upgrader mechanism needs to be built optionally. It may be easy technically, by including an empty header. Make sure this option is available at building stage: disable upgrader can be done at building stage with a simple flag.

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"


1. Check in upgrader_mobile.h and it will be replaced by code-gen from yaml file in next few pr. Let me know if this file makes sense.
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

[ghstack-poisoned]

cccclai commented:
> 2\. optionally

1. Resolve in the latest commit
2. Let me figure it about in the pr to building optionally

iseeyuan commented on torch/csrc/jit/mobile/upgrader_mobile.h:
```diff
@@ -0,0 +1,74 @@
+
+static std::vector<mobile::Function> kUpgraderFunctions(
+    {
+            mobile::Function::get(
```
Could we do something like this:
```
Code code{
{{STOREN, 1, 2}, {LOAD, 1, 0}, ...}, // instructions_
{{"aten::is_floating_point", ""}, {"aten::div", "Tensor"}, {"aten::div", "Tensor_mode"}}, //op_names_
... //constants
...//types
4 // register_size
};
```
It's pretty much direct memory initialization of function. Probably @qihqi could have suggestions on how to initialize a function directly.

It would be very expensive to use tuple and IValue and they may not be necessary.

qihqi replied:
Yes you can create a Function by using it directly: https://github.com/pytorch/pytorch/blob/38c77539e8fbff5b2245e30a0f612efc0c312cae/torch/csrc/jit/mobile/function.h

Basically it takes name as QualifiedName in constructor, then call append_instruction to add instruction and append_operator to add operator.

cccclai replied:
Address in the latest commit.

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"


1. Check in upgrader_mobile.h and it will be replaced by code-gen from yaml file in next few pr. Let me know if this file makes sense.
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and it will be replaced by code-gen from yaml file in next few pr. Let me know if this file makes sense.
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and it will be replaced by code-gen from yaml file in next few pr. Let me know if this file makes sense.
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and it will be replaced by code-gen from yaml file in next few pr. Let me know if this file makes sense.
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and it will be replaced by code-gen from yaml file in next few pr. Let me know if this file makes sense.
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

cccclai commented on torch/csrc/jit/mobile/upgrader_mobile.h:
```diff
@@ -0,0 +1,86 @@
+static std::vector<ByteCodeFunctionWithOperator> kUpgraderByteCode(
+    {ByteCodeFunctionWithOperator({
+        mobile::Function::registerFunc(
+            "div_Tensor_0_3",
```
Ohh sorry update it again.

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and it will be replaced by code-gen from yaml file in next few pr. Let me know if this file makes sense.
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and it will be replaced by code-gen from yaml file in next few pr. Let me know if this file makes sense.
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

cccclai commented on torch/csrc/jit/mobile/import.cpp:
```diff
@@ -11,6 +11,7 @@
 #include <torch/csrc/jit/api/compilation_unit.h>
 #include <torch/csrc/jit/mobile/interpreter.h>
 #include <torch/csrc/jit/mobile/observer.h>
+#include <torch/csrc/jit/mobile/upgrader_mobile.h>
```
Address in the latest commit.

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and it will be replaced by code-gen from yaml file in next few pr. Let me know if this file makes sense.
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and it will be replaced by code-gen from yaml file in next few pr. Let me know if this file makes sense.
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Use check in cpp source file for upgrader"




1. Check in upgrader_mobile.h and upgrader_mobile.cpp
2. Add test to parse all bytecode from upgrader_mobile.h

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32087295/)!

Differential Revision: [D32087295](https://our.internmc.facebook.com/intern/diff/D32087295)

[ghstack-poisoned]
